pkg/webhooks/pod_webhook.go

```diff
@@ -113,16 +113,50 @@ func (p *PodWebhook) Default(ctx context.Context, obj runtime.Object) error {
 		} else {
 			groupUniqueKey = pod.Labels[leaderworkerset.GroupUniqueHashLabelKey]
 		}
-		_, foundEpKey := pod.Annotations[leaderworkerset.ExclusiveKeyAnnotationKey]
-		if foundEpKey && !exclusiveAffinityApplied(*pod) {
-			SetExclusiveAffinities(pod, groupUniqueKey)
+		epKey, foundEpKey := pod.Annotations[leaderworkerset.ExclusiveKeyAnnotationKey]
+		if (foundEpKey) && !exclusiveAffinityApplied(*pod, epKey) {
+			SetExclusiveAffinities(pod, groupUniqueKey, epKey, leaderworkerset.GroupUniqueHashLabelKey)
+		}
+		_, foundSubGroupSize := pod.Annotations[leaderworkerset.SubGroupSizeAnnotationKey]
+		if foundSubGroupSize && pod.Labels[leaderworkerset.SubGroupIndexLabelKey] == "" {
+			// The leader pod will always lands on SubGroup 0 in the same
+			// topology regardless if it requests any accelerator resources
+			pod.Labels[leaderworkerset.SubGroupIndexLabelKey] = "0"
+			subGroupUniqueKey := genGroupUniqueKey(pod.Name, "0")
+			pod.Labels[leaderworkerset.SubGroupUniqueHashLabelKey] = subGroupUniqueKey
+			subEpKey, foundSubEpKey := pod.Annotations[leaderworkerset.SubGroupExclusiveKeyAnnotationKey]
+			if foundSubEpKey && !exclusiveAffinityApplied(*pod, subEpKey) {
+				SetExclusiveAffinities(pod, subGroupUniqueKey, subEpKey, leaderworkerset.SubGroupUniqueHashLabelKey)
+			}
 		}
 	} else {
 		_, workerIndex := statefulsetutils.GetParentNameAndOrdinal(pod.Name)
 		if workerIndex == -1 {
 			return fmt.Errorf("parsing pod ordinal for pod %s", pod.Name)
 		}
 		pod.Labels[leaderworkerset.WorkerIndexLabelKey] = fmt.Sprint(workerIndex)
+		subGroupSize, foundSubGroupSize := pod.Annotations[leaderworkerset.SubGroupSizeAnnotationKey]
+		if foundSubGroupSize && pod.Labels[leaderworkerset.SubGroupIndexLabelKey] == "" {
+			subGroupSizeInt, err := strconv.Atoi(subGroupSize)
+			if err != nil {
+				return err
+			}
+
+			var subGroupIndexKey string
+			if acceleratorutils.PodRequestsTPUs(pod.Spec) {
+				subGroupIndexKey = acceleratorutils.AddTPUSubGroupLabels(pod, workerIndex, subGroupSizeInt)
+			} else {
+				subGroupIndexKey = fmt.Sprint(workerIndex / subGroupSizeInt)
+				pod.Labels[leaderworkerset.SubGroupIndexLabelKey] = subGroupIndexKey
+			}
+			leaderName := pod.Annotations[leaderworkerset.LeaderPodNameAnnotationKey]
+			subGroupUniqueKey := genGroupUniqueKey(leaderName, subGroupIndexKey)
```

Comment:
the hash key shouldn't depend on whether or not the pod is requesting TPUs. The subgroup index is not dependent on that.

Reply:
It does to an extend. If the pod is requesting TPUs, and the leader requests TPU resources, then it changes which index goes into which subgroup, changing the `subGroupIndexKey`. I can make it more clear that it is this specific case that changes it, but since the check is specific to TPUs opted to have it inside the function `AddTPUSubGroupLabels`.

Reply:
The definition of subgroups is not related to whether or not TPUs are being requested: 

```
        // The number of pods per subgroup. This value is immutable,
	// and must not be greater than LeaderWorkerSet.Spec.Size.
	// Size must be divisible by subGroupSize in which case the
	// subgroups will be of equal size. Or size - 1 is divisible
	// by subGroupSize, in which case the leader is considered as
	// the extra pod, and will be part of the first subgroup.
```

If we want the leader to be split out from the first subgroup and make it its own subgroup, we can add an explicit knob on the API for the user to express this intent.

Reply:
Makes sense, will check the value of size instead to determine whether or not the `subGroupIndexKey` changes.

Reply:
fixed